### PR TITLE
[core] Introduce TTL for tag

### PR DESCRIPTION
**Pull request by wwj6591812**
### Purpose

Linked issue: [Introduce TTL for tag](https://github.com/apache/paimon/issues/3127)

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java [hunk 1]
```diff
@@ -27,30 +27,40 @@
...
+import java.util.Set;
 import java.util.SortedMap;
 
 import static org.apache.paimon.Snapshot.FIRST_SNAPSHOT_ID;
 import static org.apache.paimon.shade.guava30.com.google.common.base.MoreObjects.firstNonNull;
+import static org.apache.paimon.utils.Preconditions.checkArgument;
 import static org.apache.paimon.utils.Preconditions.checkState;
 
 /** A manager to create tags automatically. */
 public class TagAutoCreation {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java:52]: Only creating this class when table has auto tag creation options.

We need to create another class? Or create a `TagAutoManager` to wrap `TagAutoCreation` and `TagAutoExpire`.

File: paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java [hunk 2]
```diff
@@ -144,29 +157,66 @@ private void tryToTag(Snapshot snapshot) {
...
+        // handle auto-created and non-auto-created-tags here
+        Set<String> deleteTags = new HashSet<>();
+        SortedMap<Snapshot, List<String>> tags = tagManager.tags();
+        for (Map.Entry<Snapshot, List<String>> entry : tags.entrySet()) {
+            checkArgument(
+                    entry.getKey() instanceof Tag,
+                    "There is a bug, snapshot in tagManager.tags() must be Tag.");
+            Tag tag = (Tag) entry.getKey();
+            LocalDateTime createTime = tag.createTime();
+            Duration timeRetained = tag.timeRetained();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java:208]: So what happen if `timeRetained` is null?

**Review by JingsongLi — CHANGES_REQUESTED** [review]

**Comment by JingsongLi** [comment]
Hi @wwj6591812 , do not write issue id in PR name or commit message.

In PR description, this is enough.

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                totalRecordCount,
+                deltaRecordCount,
+                changelogRecordCount,
+                watermark,
+                statistics);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(
```
[the remark this answers is not in the document]
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java]: done

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+                timeRetained);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof Tag)) {
+            return false;
+        }
+        Tag that = (Tag) o;
+        return Objects.equals(version, that.version)
```
[the remark this answers is not in the document]
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java]: done

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+/** Snapshot with tagCreateTime and tagTimeRetained. */
+public class Tag extends Snapshot {
+
+    private static final String FIELD_CREATE_TIME = "createTime";
+    private static final String FIELD_TIME_RETAINED = "timeRetained";
+
+    @JsonProperty(FIELD_CREATE_TIME)
+    @Nullable
+    private final LocalDateTime createTime;
```
[the remark this answers is not in the document]
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java]: done

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+    @JsonProperty(FIELD_TIME_RETAINED)
+    @Nullable
+    private final Duration timeRetained;
```
[the remark this answers is not in the document]
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java]: done

File: paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java (hunk 1 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java]: done
I create a TagAutoManager to wrap TagAutoCreation and TagAutoExpire.

File: paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java (hunk 2 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java]: done
In order to compatibility old version.
New Fields tagCreateTime and tagTimeRetained in Tag should use default value when old version snapshot/tag json to new version Tag Object.

**Comment by wwj6591812** [comment]
> Hi @wwj6591812 , do not write issue id in PR name or commit message.
> 
> In PR description, this is enough.

ok. done

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    @Nullable
+    private final LocalDateTime tagCreateTime;
+
+    @JsonProperty(FIELD_TAG_TIME_RETAINED)
+    @Nullable
+    private final Duration tagTimeRetained;
+
+    @JsonCreator
+    public Tag(
+            @JsonProperty(FIELD_ID) long id,
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:60]: Why there is no version?

File: paimon-core/src/main/java/org/apache/paimon/tag/TagAutoManager.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    public void run() {
+        if (tagAutoCreation != null) {
+            tagAutoCreation.run();
+        }
+        if (tagAutoExpire != null) {
+            tagAutoExpire.run();
+        }
+    }
+
+    @Nullable
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/tag/TagAutoManager.java:51]: It is not-null.

Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/TagAutoManager.java]: done

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java (hunk 7 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:61]: added

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+/** Snapshot with tagCreateTime and tagTimeRetained. */
+public class Tag extends Snapshot {
+
+    private static final String FIELD_TAG_CREATE_TIME = "tagCreateTime";
+    private static final String FIELD_TAG_TIME_RETAINED = "tagTimeRetained";
+    private static final LocalDateTime DEFAULT_TAG_CREATE_TIME =
+            LocalDateTime.of(1970, 1, 1, 0, 0, 0, 0);
+
+    @JsonProperty(FIELD_TAG_CREATE_TIME)
+    @Nullable
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:51]: @JsonInclude(JsonInclude.Include.NON_NULL)

File: paimon-core/src/main/java/org/apache/paimon/Snapshot.java [hunk 10]
```diff
@@ -64,6 +65,7 @@
  *       there is no compatibility issue.
  * </ul>
  */
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/Snapshot.java:68]: For this: https://github.com/apache/paimon/pull/3202

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    private final LocalDateTime tagCreateTime;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:52]: This field will cause old readers to be unable to read the new format, which can be a bit troublesome. Can we just use file creation time?

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+    @JsonProperty(FIELD_TAG_TIME_RETAINED)
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:54]: @JsonInclude(JsonInclude.Include.NON_NULL)

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java (hunk 9 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:56]: done

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java (hunk 12 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:60]: done

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java (hunk 11 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:57]: I have maked auto-create-tag use Snapshot json, not new Tag json.

**Comment by wwj6591812** [comment]
I my fork Paimon，the UT all success.
paimon ut

**Comment by wwj6591812** [comment]
I my fork Paimon，the UT all success.
Please trigger this pr UT. 
paimon ut

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 13]
```diff
@@ -997,6 +999,12 @@ public class CoreOptions implements Serializable {
                     .noDefaultValue()
                     .withDescription("The maximum number of tags to retain.");
 
+    public static final ConfigOption<Duration> TAG_TIME_RETAINED =
+            key("tag.time-retained")
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:1003]: `tag.default-time-retained`

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 14]
```diff
@@ -997,6 +999,12 @@ public class CoreOptions implements Serializable {
...
+                    .durationType()
+                    .defaultValue(DEFAULT_TAG_TIME_RETAINED)
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:1005]: default none.

File: paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java [hunk 15]
```diff
@@ -482,18 +482,36 @@ public void createTag(String tagName, long fromSnapshotId) {
...
+        createTag(tagName, coreOptions().tagTimeRetained(), createTagInternal(fromSnapshotId));
+    }
+
+    @Override
+    public void createTag(String tagName, Duration timeRetained, long fromSnapshotId) {
+        createTag(tagName, timeRetained, createTagInternal(fromSnapshotId));
     }
 
     @Override
     public void createTag(String tagName) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:499]: We should be very careful, old way should have no other fields in the json.

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java (hunk 13 above)
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:1003]: Now, it is just a default value for time-retained in creating tag.

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java (hunk 14 above)
Comment (wwj6591812) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java]: done
good idea

File: paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java (hunk 15 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:500]: good, thanks very much!
I have change TAG_DEFAULT_TIME_RETAINED to default none. And In TagManager.createTag() when createTag is null then create old snapshot json.

**Pull request closed without merging**